lib/generate_letters.ml

```diff
@@ -1,42 +1,179 @@
-(** letter_deck is a list containing 7 random chars at any time. To be
-    used in a game of scrabble. If the char has been used in the game, a
-    new char will take its place, so there will never be an empty index.
-    Requires each char in the list to be an alphabetical letter at any
-    given time. *)
+
+(** [index_of_letter curr letter lst] is the index of the first
+    occurrence of this character in the list. Requires letter to be
+    present in lst and curr to be the current index (starting at 0).
+    index_of_letter : int -> char -> letter_deck -> int*)
+let rec index_of_letter curr letter = function
+  | [] -> raise (Failure "Empty List")
+  | h :: t ->
+      if h = letter then curr else index_of_letter (curr + 1) letter t
+
+(* vowel_amounts is an association list that links each vowel to the
+   probability of drawing that vowel given it's frequency in the tiles*)
+let vowel_amounts : (float * char list) list =
+  [
+    (12.1, [ 'U' ]);
+    (24.2, [ 'O' ]);
+    (27.2, [ 'A'; 'I' ]);
+    (36.3, [ 'E' ]);
+  ]
+
+(* consonant_amounts is an association list that links each consonant to
+   the probability of drawing that consonant given it's frequency in the
+   tiles*)
+let consonant_amounts : (float * char list) list =
+  [
+    (6.2, [ 'Q'; 'Z'; 'J'; 'X'; 'K' ]);
+    (12.5, [ 'F'; 'H'; 'V'; 'W'; 'Y'; 'B'; 'C'; 'M'; 'P' ]);
+    (18.7, [ 'G' ]);
+    (25., [ 'D'; 'S'; 'L' ]);
+    (37.5, [ 'T'; 'R'; 'N' ]);
+  ]
+
+(* alias table reference https://www.keithschwarz.com/darts-dice-coins/*)
+(* probability list needed for an alias table*)
+let prob = [ 0.42; 0.92; 0.56; 1.; 0.42; 0.06 ]
+
+(* alias list needed for an alias table*)
+let alias = [ 1; 2; 3; 0; 2; 2 ]
+
+(** [biased_bool prob total] is true for heads and false for tails. This
+    method simulates a biased coin, with a [prob]/[total] chance of
+    being heads. Requires [prob] <= [total].*)
+let biased_bool prob total =
+  let pick = Random.int total in
+  if pick < prob then true else false
+
+(** [index_item ind curr lst] is the item of lst at index ind. The
+    parameter curr is the current index.*)
+let rec index_item curr ind = function
+  | [] -> raise Not_found
+  | h :: t -> if curr = ind then h else index_item (curr + 1) ind t
+
+(** [get_random_item lst] is a random element from lst. *)
+let get_random_item lst =
+  let ind = Random.int (List.length lst) in
+  index_item 0 ind lst
+
+(** [inner_match_probability prob accum lst] is a helper for
+    [match_probability]. Requires [accum] = 0*)
+let rec inner_match_probability prob accum lst =
+  match lst with
+  | [] -> []
+  | (p, ls) :: t ->
+      let curr_prob = accum +. p in
+      if prob <= curr_prob then ls
+      else inner_match_probability prob curr_prob t
+
+(** [match_probability prob lst] is the list of values in [lst] given
+    probability [prob]. Requires [prob] < 100 and [lst] to be a (float *
+    char list) association list representing (probability * value list)
+    tuples.*)
```

Comment:
I don't understand this comment / I don't understand what this function does

Reply:
idrk how to explain it better its just how im determining which letter to use given a random number based on its probability. can explain more in person